errtracker/errtracker.go

```diff
@@ -67,6 +69,80 @@ var (
 	timeNow  = time.Now
 )
 
+type reportsDB struct {
+	fname string
+
+	// map of hash(dupsig) -> time-of-report
+	reported map[string]time.Time
+
+	cleanupTime time.Duration
+}
+
+func dupSigHash(s string) string {
+	h := sha512.New()
+	io.WriteString(h, s)
+	return fmt.Sprintf("%x", h.Sum(nil))
+}
+
+func newReportsDB(fname string) (*reportsDB, error) {
+	db := &reportsDB{
+		fname:       fname,
+		reported:    make(map[string]time.Time),
+		cleanupTime: time.Duration(7 * 24 * time.Hour),
+	}
+
+	f, err := os.Open(fname)
+	if err != nil {
+		if os.IsNotExist(err) {
+			return db, nil
+		}
+		return nil, err
+	}
+	defer f.Close()
+
+	dec := json.NewDecoder(f)
+	if err := dec.Decode(&db.reported); err != nil {
+		// be robust here, if the data is corrupted for whatever
+		// reason we just create a fresh DB
+		return db, nil
+	}
+
+	return db, nil
+}
+
+func (db *reportsDB) AlreadyReported(dupSig string) bool {
+	_, reported := db.reported[dupSigHash(dupSig)]
+	return reported
+}
+
+func (db *reportsDB) cleanupOldRecords() {
+	now := time.Now()
+	for dupSigHash, reportTime := range db.reported {
+		if now.After(reportTime.Add(db.cleanupTime)) {
+			delete(db.reported, dupSigHash)
+		}
+	}
+}
+
+func (db *reportsDB) MarkReported(dupSig string) error {
```

Comment:
`// MarkReported marks an error report as reported to the error tracker`

The same suggestion to take the report instead of the hash.